cardano-cli/src/Cardano/CLI/EraBased/Run/Transaction.hs

```diff
@@ -1337,28 +1338,37 @@ createTxMintValue era (val, scriptWitnesses) =
       caseShelleyToAllegraOrMaryEraOnwards
         (const (txFeatureMismatchPure (toCardanoEra era) TxFeatureMintValue))
         ( \w -> do
-            -- The set of policy ids for which we need witnesses:
-            let witnessesNeededSet :: Set PolicyId
-                witnessesNeededSet =
-                  fromList [pid | (AssetId pid _, _) <- toList val]
+            let policiesWithAssets :: [(PolicyId, AssetName, Quantity)]
+                policiesWithAssets = [(pid, assetName, quantity) | (AssetId pid assetName, quantity) <- toList val]
 
-            let witnessesProvidedMap :: Map PolicyId (ScriptWitness WitCtxMint era)
+                -- The set of policy ids for which we need witnesses:
+                witnessesNeededSet :: Set PolicyId
+                witnessesNeededSet = fromList [pid | (pid, _, _) <- policiesWithAssets]
+
+                witnessesProvidedMap :: Map PolicyId (ScriptWitness WitCtxMint era)
                 witnessesProvidedMap = fromList $ gatherMintingWitnesses scriptWitnesses
                 witnessesProvidedSet = Map.keysSet witnessesProvidedMap
 
-            -- Check not too many, nor too few:
+                policiesWithWitnesses =
+                  Map.fromListWith
+                    (<>)
+                    [ (pid, [(assetName, quantity, BuildTxWith witness)])
+                    | (pid, assetName, quantity) <- policiesWithAssets
+                    , witness <- maybeToList $ Map.lookup pid witnessesProvidedMap
+                    ]
+
             validateAllWitnessesProvided witnessesNeededSet witnessesProvidedSet
             validateNoUnnecessaryWitnesses witnessesNeededSet witnessesProvidedSet
-            return (TxMintValue w val (BuildTxWith witnessesProvidedMap))
+            pure $ TxMintValue w policiesWithWitnesses
         )
         era
  where
   gatherMintingWitnesses
-    :: [ScriptWitness WitCtxMint era]
+    :: [(Maybe PolicyId, ScriptWitness WitCtxMint era)]
     -> [(PolicyId, ScriptWitness WitCtxMint era)]
   gatherMintingWitnesses [] = []
-  gatherMintingWitnesses (sWit : rest) =
-    case scriptWitnessPolicyId sWit of
+  gatherMintingWitnesses ((mPid, sWit) : rest) =
```

Comment:
We need a new data definition.

`fetchReferenceMintingScriptWitness :: UTxO era -> ScriptWitness WitCtxMint era -> UpdatedReferenceScriptWitness`

and in the case where the user chooses to supply it 

`updateReferenceScriptWitness :: PolicyId -> ScriptWitness WitCtxMint era -> UpdatedReferenceScriptWitness`

Propagating `Maybe PolicyId` may work but it results in poor readability. If the user does not supply the policy id  we really want is the ability to query the UTxO to get the reference script witnesses so we can calculate the policy ID.

Reply:
This would be the short to medium term fix until we can figure out how to encapsulate this in `cardano-api`.

Reply:
I agree about the first case with `fetchReferenceMintingScriptWitness`. I'd do this in a follow-up PR.

What is not clear to me is how to thread in `Maybe PolicyId` from the CLI argument parser to this place. What's your idea about this scenario, when the user provides `Maybe PolicyId` manually?

Reply:
I would start with something like: `data MintingReferenceScriptWitness era = MintingReferenceScriptWitness PolicyId (ScriptWitness WitCtxMint era)`

> I'd do this in a follow-up PR.

In this instance I would insist its done in this PR.

Reply:
@Jimbo4350 I'm sorry, I'm still not seeing it. The current solution relies on treating minting witnesses on equal with other witnesses. This allows us to have a list of `ScriptWitness ctx era` and process it uniformly (read from file, add to transaction etc). I'm not seeing how this new datatype would fit in?

Reply:
No worries, let me push some changes tomorrow to make sure what I am suggesting can actually work.

Reply:
I've removed my commend as I'm still digesting the idea and experimenting with it. Feel free to push.